### PR TITLE
Implement port mapping

### LINKED ISSUE #938: Implement port mapping for containers
This would enable the docker use case where containers map ports through the docker server. This can be done by implementing the correct NAT rules on the VCH and container side. This would also mean assigning an IP address to the VCH on the relevant bridge network(s).

Acceptance:
* VCH has an IP address on the relevant bridge network(s)
* `docker create -p/-P` works

### PR DESCRIPTION
**Pull request by hmahmood**
Fixes #938

### CODE REVIEW

**Comment by jooskim** [comment]
Do we now have port mapping info available to UI? Or can I access port mapping info through mob interface?

File: isos/appliance/nat-setup [hunk 1]
```diff
@@ -58,15 +56,30 @@ iptables-restore <<RULES
...
+-A INPUT -p tcp -m tcp --dport 22 -j ACCEPT
+-A INPUT -p icmp -m icmp --icmp-type 8 -j ACCEPT
+-A INPUT -p udp -m udp --dport 68 -j ACCEPT
+-A INPUT -p tcp -m tcp --dport 2378 -j ACCEPT
+-A INPUT -p tcp -m tcp --dport 2376 -j ACCEPT
+-A INPUT -i bridge -p udp -m udp --dport 53 -j ACCEPT
+-A INPUT -i bridge -p tcp -m tcp --dport 53 -j ACCEPT
+-A INPUT -p tcp -m tcp --dport 8080 -j ACCEPT
+-A INPUT -p udp -j REJECT --reject-with icmp-port-unreachable
+-A INPUT -p tcp -j REJECT --reject-with tcp-reset
```
Comment (andrewtchin) [inline isos/appliance/nat-setup]: 👏

File: lib/apiservers/engine/backends/portmap/portmap.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+type Operation int
+
+const (
+	Map Operation = iota
+	Unmap
+)
+
+type PortMapper interface {
+	MapPort(op Operation, ip net.IP, port int, proto string, destIP string, destPort int, strcIface, destIface string) error
```
Comment (andrewtchin) [inline lib/apiservers/engine/backends/portmap/portmap.go]: is `strcIface` a typo?

**Comment by hmahmood** [comment]
@jooskim port mapping info is not available in the guestinfo, only the exposed ports are. So, for instance, if a user specifies `-p 8080:80` during `docker create/run`, only `80` would be available in guestinfo. The command `docker ports` will give you the full info though, and I think that is still under development here #854

**Comment by jooskim** [comment]
@hmahmood Ok thanks. This is just out of curiosity, but what happens if more than one port is exposed to host? like 80 and 443. Do we get both ports in guestinfo?

**Comment by hmahmood** [comment]
@jooskim yes you will get both ports in guestinfo.

File: lib/apiservers/engine/backends/portmap/portmap.go (hunk 2 above)
Comment (hmahmood) [inline lib/apiservers/engine/backends/portmap/portmap.go]: Yes, I thought I had fixed that :-/

File: Makefile [hunk 3]
```diff
@@ -282,7 +282,8 @@ $(iso-base): isos/base.sh isos/base/*.repo isos/base/isolinux/** isos/base/xorri
 	@$(TIME) $< -c $(BIN)/yum-cache.tgz -p $@
 
 # appliance staging - allows for caching of package install
-$(appliance-staging): isos/appliance-staging.sh $(iso-base)
+$(appliance-staging): isos/appliance-staging.sh $(i
+so-base)
```
Comment (caglar10ur) [inline Makefile]: I think this commit escaped from a jail :)

File: isos/appliance/nat-setup [hunk 4]
```diff
@@ -43,11 +43,9 @@ echo 2 > /proc/sys/net/ipv4/ip_dynaddr
 # too many routers are still ignorant
 echo 0 > /proc/sys/net/ipv4/tcp_ecn
 
-# wait for the bridge interface to exist - 100ms wait
-until ip link show dev bridge;do sleep 0.1;done
-
 # blow away any existing rules with iptables-restore
 iptables-restore <<RULES
+# Generated by iptables-save v1.6.0 on Tue Jul 19 22:57:40 2016
```
Comment (caglar10ur) [inline isos/appliance/nat-setup]: do we care about those headers ^v?

**Comment by caglar10ur** [comment]
LGTM :sheep: :it:

**Comment by andrewtchin** [comment]
lgtm

**Comment by hmahmood** [comment]
:recycle:

**Pull request merged**